### ISSUE TITLE
Sorbet does not always infer the type of global variables correctly

### ISSUE THREAD

**Issue by manhhung741** [sorbet/sorbet#1465]
#### Input

[→ View on sorbet.run](TODO_replace_this_with_your_sorbet_run_link)

See full example: https://gist.github.com/jrunning/fe882c32580e33c6aac436ec97453ee4

```ruby
# -- in rollout_setup.rb
$rollout = Rollout.new(::CachedRolloutStore.new)
T.reveal_type($rollout) # got Rollout

# -- in foo.rb
T.reveal_type($rollout) # got T.untyped
```

#### Observed output

`$rollout` is `T.untyped` in a different file.
`$rollout.method_call` is not type checked
but when passing `$rollout` as a param, its type is inferred correctly

#### Expected behavior

`$rollout` should have correct type & its method should be typechecked everywhere

- - -

See the original post on StackOverFlow: https://stackoverflow.com/questions/57313647/how-do-i-specify-the-type-of-global-variable-in-sorbet?noredirect=1#comment101120657_57313647

**Comment by DarkDimius** [comment]
This is by design. If you'd like to reliably have it figured out, please declare globals with T.let
```
$rollout = T.let(Rollout.new(::CachedRolloutStore.new), Rollout)
```

This indicates that you want this variable to have this type, rather than it happens to have been assigned to this type in this callsite.

**Comment by DarkDimius** [comment]
Note that in `# typed: strict` that^^^ is required

**Comment by manhhung741** [comment]
Even with T.let, it doesn’t detect the object’s type correctly on a
different file.

**Comment by DarkDimius** [comment]
Ack, then that's a bug. Thank you!

**Comment by manhhung741** [comment]
Yep. You can find in the attached gist that there is a repro with `T.let`.
That gist can be used as a test case.
